miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/CachedMiniAppVerifier.kt

```diff
@@ -0,0 +1,88 @@
+package com.rakuten.tech.mobile.miniapp.storage
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.util.Base64
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import androidx.security.crypto.EncryptedSharedPreferences
+import androidx.security.crypto.MasterKeys
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.io.File
+import java.io.IOException
+import java.security.MessageDigest
+import kotlin.coroutines.CoroutineContext
+
+internal class CachedMiniAppVerifier
+@VisibleForTesting constructor(
+    private val prefs: SharedPreferences,
+    private val coroutineDispatcher: CoroutineDispatcher
+) : CoroutineScope {
+
+    constructor(context: Context) : this(
+        prefs = EncryptedSharedPreferences.create(
+            "com.rakuten.tech.mobile.miniapp.cache.hash",
+            MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
+            context,
+            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
+            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
+        ),
+        coroutineDispatcher = Dispatchers.IO
+    )
+
+    override val coroutineContext: CoroutineContext
+        get() = coroutineDispatcher
+
+    /** Verifies that the cached files for the Mini App have not been modified. */
+    fun verify(appId: String, directory: File): Boolean {
+        return try {
+            val hash = calculateHash(directory)
+            val storedHash = prefs.getString(appId, null) ?: ""
+            hash == storedHash
+        } catch (e: IOException) {
+            Log.e(TAG, "Failed to verify the stored Mini App hash.", e)
+            false
+        }
+    }
+
+    /** Stores hash in encrypted shared preferences. This runs asynchronously so it will return immediately. */
+    fun storeHash(appId: String, directory: File) {
+        launch {
+            try {
+                val hash = calculateHash(directory)
+                prefs.edit().putString(appId, hash).apply()
+            } catch (e: IOException) {
+                Log.e(TAG, "Failed to store hash for the Mini App.", e)
+            }
+        }
+    }
+
+    private fun calculateHash(
+        directory: File
+    ): String {
```

Comment:
Also, we can catch the `IOException` in here instead of `verify(...)` and `storeHash(...)` function. My total idea would be -

```
private fun calculateHash(
    directory: File
): String = runBlocking {
        try {
             // calculating hash
        } catch (e: IOException) {
            throw MiniAppSdkException(e)
        }
 }
```

Reply:
I've moved the `Exception` into here instead. But I don't think using `runBlocking` is the right solution here. We want the `verify` function to run synchronously but the `storeHash` function should run asynchronously. 

However, I realized that we don't need to create a new CoroutineScope here - instead I've changed `storeHash` to a `suspend` function and run the coroutine using `async`. What do you think?

Reply:
> We want the verify function to run synchronously but the storeHash function should run asynchronously.

I think, it sounds good 👍